### PR TITLE
Improved graphs of stats page 

### PR DESCRIPTION
**Pull request by Souravirus**
fixing issue #2068 
Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [X] code is in uniquely-named feature branch and has no merge conflicts
* [X] PR is descriptively titled
* [X] PR body includes `fixes #0000`-style reference to original issue #
* [X] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software

We have a loose schedule of reviewing and pulling in changes every Tuesday and Friday, and publishing changes on Fridays. 

Thanks!

### CODE REVIEW

**Comment by Souravirus** [comment]
@jywarren @ebarry @sagarpreet-chadha Check out these graphs I have added to stats page 
A normal pic:
screenshot from 2018-02-27 03-09-27
When we hover over the graph:
screenshot from 2018-02-27 03-09-46
Graph in small screens:
screenshot from 2018-02-27 03-10-31

**Comment by jywarren** [comment]
Oh cool! These look much nicer than the profile pages on narrow displays. Do graphs work better on narrow screens than heat maps? 

Fantastic work here! Here too, could we just show month names like `Jan` `Feb` ... ? That'll make the display a bit simpler and more readable.

You're on fire these days!!! 👍 👍 🎈

**Comment by Souravirus** [comment]
Yeah these graphs are better designed for small screens than heat maps.  Okk I will show month names instead of week numbers. Thank you

**Comment by Souravirus** [comment]
@jywarren Check out these improved graphs with monthly contributions being shown
Normal Pic
screenshot from 2018-02-27 18-49-56
When hovered over the graph
screenshot from 2018-02-27 18-50-08
graphs in small screen
screenshot from 2018-02-27 18-50-15

**Comment by ebarry** [comment]
Wow this looks awesome! 

For the Y axis, could we show only integers? This will save us from the dilemma of trying to figure out what "0.25 of a comment" might be.  ;)

Does clicking on the graph lead to https://publiclab.org/stats ?

**Comment by Souravirus** [comment]
@ebarry Yeah I would try removing decimal values on the Y-axis. No clicking on graph does not leads to any link. Only hovering on the graph shows us data about that particular X-axis and Y-axis value.

**Comment by ebarry** [comment]
Thank you!!! Excited to see the change in the Y-axis increments. 

I think in the original issue I had mentioned adding a link to /stats but perhaps it is not important.

**Comment by Souravirus** [comment]
@ebarry Actually there is no way to show only integer value on Y-axis as the developers of [frappe-charts](https://github.com/frappe/charts) have not given any option to do so. They actually automatically give the values in y-axis as can be seen in this code snippet:
screenshot from 2018-02-27 23-53-40
As you can see if the range at Y-axis goes below 2 they divide the interval into 4 parts. So for now I cannot do anything about the decimal values. But I have put an [issue](https://github.com/frappe/charts/issues/127) for it in their repository. So if they address the issue soon, I can implement the graph as you are saying.

**Comment by jywarren** [comment]
Would it be possible to actually show the weekly tallies, but to have labels shown for only months? This is really amazing.

**Comment by ebarry** [comment]
Wow thanks for checking into this! This all looks super!

**Comment by Souravirus** [comment]
No @jywarren with frappe-charts it is not possible because they have fixed X-axis values and their respective Y-axis values. So I cannot show this with frappe-charts.

**Comment by Souravirus** [comment]
Hey @jywarren I tried what you told yesterday by giving blank values to some weeks and giving the middle week of a month, the name of the month. But there is a problem here, since the month name are three letter words.So some month names are missing and the graph comes out something like this:
screenshot from 2018-03-01 15-17-04
Now when I named the months somewhat shorter it came out some like this:
screenshot from 2018-03-01 15-19-16
Now another option is naming months by their numbers, So it would appear something like this:
screenshot from 2018-03-01 15-20-58
So what sould I do?

**Comment by jywarren** [comment]
Ha, very clever solution! And this clipping is just in CSS? I wonder if we could "inspect" the css and see why it's clipping. Maybe there's a `overflow:hidden` style we could override? 

Very clever!!!

**Comment by Souravirus** [comment]
@jywarren this clipping is being done by an option given by frappe-charts know as is_series to show the graph in small screens. And if I remove it I can't even see the months. Instead of months a '...' sign comes.  
Something like this:
screenshot from 2018-03-03 21-42-40

**Comment by jywarren** [comment]
OK -- what about if we just used a single letter for each month. It's not ideal, but we can refine this more later. Perhaps even by contributing upstream to frappe charts. Thanks, with this last change this should be good to go. 

Could we also have the graphs be solid colors as in the current bar charts? I think the gradients are not perfect, although I'm open to trying out different solid colors. Thank you very much for your hard work on this!

**Comment by Souravirus** [comment]
@jywarren I will make months in one letter and I would change the color to solid.

**Comment by Souravirus** [comment]
@jywarren, I can't remove the gradient either I have to keep the region fill or not. Something like this,
screenshot from 2018-03-12 21-51-58
Or else I can introduce bar graph which has solid colors and change colors of the bar if you tell something like this:
screenshot from 2018-03-12 21-58-11
Now tell me what should I do?

**Comment by jywarren** [comment]
Hmm, ok. Let's do the bars. Sorry i was away for a while, this is almost there!! Very much appreciate your hard work on it 👍 👍

**Comment by Souravirus** [comment]
Okk I will keep the bar graphs. Thanks for your suggestion!!

**Comment by Souravirus** [comment]
@jywarren, I have added bar graphs on stats page. Please review it. Thank you!!

**Comment by jywarren** [comment]
Super. @publiclab/reviewers, is someone available to review this? Thanks!!!!

**Comment by jywarren** [comment]
I think for one thing, do you think you could go through and make the indentation really consistent? Sorry to ask, but it really does help us keep the code tidy, long term!

Maybe @namangupta01 or @ryzokuken could help with some tips with RuboCop, but just manually indenting with two-space convention goes a long way!

Thank you!!!

**Comment by Souravirus** [comment]
Oh I am really sorry, I didn't use the correct indentation in some files. I would try to  use rubocop for indentation. Thank you for the suggestion @jywarren

File: test/functional/stats_controller_test.rb [hunk 1]
```diff
@@ -1,11 +1,24 @@
...
+      Node.create!(type: 'note', title:'blah', uid: 1, status: 1)
+      get :index
+      assert_equal assigns(:graph_notes), Node.contribution_graph_making('note', 52, Time.now).to_a.to_json
+  end
+  test 'should assign correct value to graph_wikis on GET stats' do
+      Node.delete_all
+      Node.create(type: 'note', title:'blah', uid: 1, status: 1)
+      Node.create(type: 'page', title:'blahblah', uid: 1, status: 1)
+      get :index
+      assert_equal assigns(:graph_wikis), Node.contribution_graph_making('page', 52, Time.now()).to_a.to_json
```
Comment (Gauravano) [inline test/functional/stats_controller_test.rb]: perfect, just spacing between methods needed.

File: app/controllers/stats_controller.rb [hunk 2]
```diff
@@ -66,9 +66,10 @@ def index
       .where(timestamp: @time.to_i - 1.years.to_i..@time.to_i)
       .count / 52.0
 
-    @graph_notes = Node.weekly_tallies('note', 52, @time).to_a.sort.to_json
-    @graph_wikis = Node.weekly_tallies('page', 52, @time).to_a.sort.to_json
-    @graph_comments = Comment.comment_weekly_tallies(52, @time).to_a.sort.to_json
+    @graph_notes = Node.contribution_graph_making('note', 52, @time).to_a.to_json
+    @graph_wikis = Node.contribution_graph_making('page', 52, @time).to_a.to_json
+    @graph_comments = Comment.contribution_graph_making(52, @time).to_a.to_json
```
Comment (Gauravano) [inline app/controllers/stats_controller.rb]: seems good

File: app/views/stats/index.html.erb [hunk 3]
```diff
@@ -47,83 +47,125 @@
...
+            show_dots: 0,
+            heatline: 1,
+            height: 115,
+            region_fill: 0, 
+            data: data[i],
+            x_axis_mode: "tick",
+            y_axis_mode: "span",
+            is_series: 1,
+            colors: ['purple']
+        });
```
Comment (Gauravano) [inline app/views/stats/index.html.erb]: seems fine

**Comment by Souravirus** [comment]
Yeah sure make it today @Gauravano

**Comment by Souravirus** [comment]
@jywarren @Gauravano Now check the indentation and spacing

File: app/models/comment.rb [hunk 4]
```diff
@@ -33,6 +33,39 @@ def self.comment_weekly_tallies(span = 52, time = Time.now)
...
+          end
+        end
+        month = month.to_i
+        #Now fetching comments per week
+        currWeek = Comment.select(:timestamp)
+                        .where(timestamp: time.to_i - week.weeks.to_i..time.to_i - (week - 1).weeks.to_i)
+                        .count
+        weeks[count] = [month, currWeek]
+        count+=1
+        week-=1
```
Comment (Gauravano) [inline app/models/comment.rb]: if possible plz add space around operators here and in node.rb too , in all other places it's good.Thanks.

Comment (Souravirus) [reply app/models/comment.rb]: Yeah I will do it. Thank you very much for the reviews !!

**Comment by Souravirus** [comment]
Hey, @jywarren @Gauravano please check it I guess its perfect from the angle of indentation and spacing.

**Comment by jywarren** [comment]
Awesome! @ebarry check it out - will publish in a day or two. Thanks everyone!!!! :+1:

**Comment by Souravirus** [comment]
Thanks very much @jywarren Hope you like it  when it gets published

**Pull request merged**